substrate/primitives/api/proc-macro/src/impl_runtime_apis.rs

```diff
@@ -787,17 +789,55 @@ fn generate_runtime_api_versions(impls: &[ItemImpl]) -> Result<TokenStream> {
 	))
 }
 
+/// replaces `Self` with explicit `ItemImpl.self_ty`.
+struct ReplaceSelfImpl {
+	self_ty: Box<Type>,
+}
+
+impl ReplaceSelfImpl {
+	/// Replace `Self` with `ItemImpl.self_ty`
+	fn replace(&mut self, trait_: &mut ItemImpl) {
+		visit_mut::visit_item_impl_mut(self, trait_)
+	}
+}
+
+impl VisitMut for ReplaceSelfImpl {
+	fn visit_generic_argument_mut(&mut self, i: &mut syn::GenericArgument) {
+		let typ = match i {
+			GenericArgument::Type(Type::Path(p))
+				if p.path.get_ident().is_some_and(|ident| ident == "Self") =>
+				Some(GenericArgument::Type(*self.self_ty.clone())),
+			_ => None,
+		};
+		if let Some(typ) = typ {
+			*i = typ;
+		}
+	}
+}
+
+/// Rename `Self` to `ItemImpl.self_ty` in all items.
+fn rename_self_in_trait_impls(impls: &mut [ItemImpl]) -> Result<()> {
+	impls.iter_mut().for_each(|i| {
+		let mut checker = ReplaceSelfImpl { self_ty: i.self_ty.clone() };
+		checker.replace(i);
+	});
+
+	Ok(())
+}
```

Comment:
The comment isn't really accurate as we only rename `Self` to `ItemImpl.self_ty` in generic arguments.

I don't know the rest of this code but wonder whether there is any harm in being more general ie replacing `Self` whereever it's encountered and not just in generic args?

ie something like (untested):

```rust
impl VisitMut for ReplaceSelfImpl {
	fn visit_type_mut(&mut self, t: &mut syn::Type) {
        let syn::Type::Path(p) = t else { return };

        if p.path.get_ident().is_some_and(|ident| ident == "Self") {
            *t = self.self_ty.clone()
        }
	}
}
```

If not, as a nit I think we could tidy the exzisting code slightly to something like:

```rust
impl VisitMut for ReplaceSelfImpl {
	fn visit_generic_argument_mut(&mut self, i: &mut syn::GenericArgument) {
        let GenericArgument::Type(Type::Path(p)) else { return };

        if if p.path.get_ident().is_some_and(|ident| ident == "Self") {
            *i = GenericArgument::Type(*self.self_ty.clone());
        }
	}
}
```

As long as the borrow checker is happy, at least :)

Reply:
applied 
```rust
impl VisitMut for ReplaceSelfImpl {
	fn visit_type_mut(&mut self, t: &mut syn::Type) {
        let syn::Type::Path(p) = t else { return };

        if p.path.get_ident().is_some_and(|ident| ident == "Self") {
            *t = self.self_ty.clone()
        }
	}
}
```